microprofile/WFLY-12313_MicroProfile_OpenAPI_1.1.adoc

```diff
@@ -0,0 +1,97 @@
+= Integrate MicroProfile OpenAPI 1.1
+:author:            Paul Ferraro
+:email:             paul.ferraro@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+This enhancement proposes to add support for https://github.com/eclipse/microprofile-open-api/tree/1.1.2[Eclipse MicroProfile OpenAPI 1.1] to WildFly by integrating the implementation from https://github.com/smallrye/smallrye-open-api[SmallRye].
+This spec provides a unified Java API for producing documentation that complies with the https://github.com/OAI/OpenAPI-Specification/blob/3.0.2/versions/3.0.2.md[OpenAPI v3 specification].
+
+An MP OpenAPI subsystem exposes an HTTP GET `/openapi` endpoint that returns an OpenAPI v3 compliant document (in YAML or JSON format) containing documention of the RESTful services of an application.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-12313
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:fburzigo@redhat.com[Fabio Burzigotti]
+
+=== Testing By
+
+[ ] Engineering
+
+[X] QE
+
+=== Affected Projects or Components
+
+The SmallRye implementation of the MicroProfile Open API specification is already integrated by the following projects:
+
+* https://thorntail.io/[Thorntail] via the https://github.com/thorntail/thorntail/tree/master/fractions/microprofile/microprofile-openapi[microprofile-openapi fraction].
+* https://quarkus.io/[Quarkus] via the https://github.com/quarkusio/quarkus/tree/master/extensions/smallrye-openapi[smallrye-openapi extension].
+
+These integrations will heavily inform, though not dictate, the corresponding integration into WildFly.
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* The microprofile-openapi spec module must be exposed to deployments
+* The `/openapi` endpoint shall be registered with the Undertow server/host associated with a deployment.
```

Comment:
It needs to be one `/openapi` per deployment.

There is currently no mechanism in the MP OpenAPI spec for aggregating multiple documents to provide a single `/openapi` for multiple deployments

Reply:
> It needs to be one /openapi per deployment.

@kenfinnigan Thank you for clarifying.

In the case where a second application is deployed to the same virtual host as an existing /openapi producing application, we can either:

1. Fail the second deployment
2. Allow the second deployment, but skip document generation process and log a warning

Reply:
I'm not a virtual host expert, but if you have two deployments on a single virtual host would that not mean you have endpoints of `vhost/dep1/openapi` and `vhost/dep2/openapi`?

Reply:
@kenfinnigan According to https://github.com/eclipse/microprofile-open-api/blob/mp-openapi-1.1.1/spec/src/main/asciidoc/microprofile-openapi-spec.adoc#51-overview

> A fully processed OpenAPI document must be available at the root URL `/openapi`.

This would seem to preclude:
* A deployment with a context path of "/openapi"
* Multiple deployments with distinct context paths sharing the same virtual host, as the locations of their `/openapi` documents would collide.

Did I misunderstand something?

Reply:
I think we need to get the spec clarified, as MP Health talks about the endpoint being off the context root of the deployment (https://download.eclipse.org/microprofile/microprofile-health-2.1/microprofile-health-spec.html#_rationale)

I believe the typical understanding is that if you have an application context root of `myapp`, then the endpoint is available at `/myapp/openapi` and not `/openapi`

Reply:
@kenfinnigan The MP Open API TCK deploys multiple applications and verifies the OpenAPI document available at the root of the ${test.url}, _irrespective_ of the deployment's default context path.

Additionally, both Thorntail and Quarkus both register the /openapi endpoint at the root of the host, without regard for any configured thorntail.context.path or quarkus.servlet.context-path, respectively.

Reply:
Thorntail and Quarkus working that way are by virtue of them not having more than one application present in a deployment.

Which TCK test deploys two JAX-RS applications into a single deployment?

Might be worth getting @EricWittmann thoughts on this too

Reply:
> Thorntail and Quarkus working that way are by virtue of them not having more than one application present in a deployment.

That's the point. Wildfly is not assuming the same "one-application" scenario, if I don't go wrong. This is the reason why I've been asking for sharing more detail about this first requirement. Just to have an idea, after reading through the above discussion, it is not clear yet "how many" `/openapi` endpoints there should be, at which level and how would they behave with respect to producing the OpenAPI document in different use cases.

Reply:
@kenfinnigan 
> Thorntail and Quarkus working that way are by virtue of them not having more than one application present in a deployment.

I understand that they only support a single application/deployment.  However, whether the Open API endpoint should be relative to the host or to the context path is _still_ relevant to the spec compliance accuracy of both frameworks, regardless of how many applications/deployments the framework supports.

> I believe the typical understanding is that if you have an application context root of myapp, then the endpoint is available at /myapp/openapi and not /openapi

If the above was true, and Thorntail's implementation is correct, then an application deployed to ThornTail using `thorntail.context.path=/mycontext` should have OpenAPI documentation available at: `host:port/mycontext/openapi`  However, this path returns a 404.  Rather, the document is only available at `host:port/openapi`.

The same holds true for Quarkus when using `quarkus.servlet.context-path=/mycontext`

> Which TCK test deploys two JAX-RS applications into a single deployment?

Sorry if I was unclear.  Let me elaborate.

The TCK runs a set of tests against a number of applications, including `airlines.war`, `airlinesReader.war`, `petstore.war`, etc.

Thorntail and Quarkus will, by default, deploy these applications to the root context.  Thus when the TCK looks for the Open API documentation at `host:port/openapi` there are no problems.

However, most multi-deployment servlet containers (including WildFly, Tomcat, etc.) will, by default (i.e. in the absence of additional meta data), deploy these applications using context paths corresponding to the name of the archive.
In WildFly, this means deploying to: `host:port/airlines.war`, `host:port/airlinesReader.war`, `host:port/petstore.war`, etc.

However, the TCK looks for the Open API documentation at ${test.url}/openapi, which is the same for each test application.  While we can certainly add WildFly-specific hooks to the test execution that appends requisite meta data necessary to change the context path of the test applications to use the root context - to me, the wording of the spec, as well as the behavior described by the TCK, suggests that the `/openapi` endpoint should be relative to the root of the host, not the context path.  For optimal service discoverability, it also makes sense to me that the `/openapi` documentation resides at a predictable location, without requiring foreknowledge of context paths.

@EricWittmann Could you help clarify this?

Reply:
My understanding of the spec is that the generated openapi doc should always be at `/openapi` regardless of how many applications are deployed to the platform.  The expectation in the spec is that all of the generated openapi docs would be merged (union).  The relevant section of the spec is here:

https://github.com/eclipse/microprofile-open-api/blob/master/spec/src/main/asciidoc/microprofile-openapi-spec.adoc#55-multiple-applications

I was always under the impression that Microprofile wasn't intended for classic app servers like Wildfly, but there it is. :)

I'm not sure the smallrye implementation supports this merging of multiple app/context OpenAPI docs very well (or at all).

Reply:
Ok, so always having worked with single deployment runtimes I've clearly not understood the spec!

Thanks for the clarification @EricWittmann

Reply:
> My understanding of the spec is that the generated openapi doc should always be at /openapi regardless of how many applications are deployed to the platform. The expectation in the spec is that all of the generated openapi docs would be merged (union). The relevant section of the spec is here:

Indeed, as stated previously, IIUC and in my opinion, the `hard` requirement no. 1 definition needs to include what at `nice-to-have` requirement no. 1 (multiple deployments), because Wildfly - as a multi deployment servlet container/application server etc. - typically allows for this to happen - also, as said previously, let's consider `ear` deployments, too...

> I was always under the impression that Microprofile wasn't intended for classic app servers like Wildfly, but there it is. :)

Probably that's the whole point: _microprofile_ vs (yet fast & beautiful) _application server_ . For this very difference - consisting in what is the expected (from the user perspective) deployment and runtime execution model between - let's say - something like Thorntail and Wildfly - main use cases and feature behavior should be described right in this proposal to clarify how Wildfly MP OpenAPI implementation will work, IMO.

@pferraro what do you think about the above suggestions? See also what you commented about usual naming strategy that Wildfly uses for app context. 
IMHO there's really need to make clear: 
* what the endpoint URL will be (consider virtual hosts definition/scope here)
* how this will work in case there are multiple deployments
* what the contents of the generated OpenAPI document will be with respect to the previously mentioned scenarios. 
Simple examples would definitely help here.

Reply:
> what the endpoint URL will be

This is mandated by the spec.  See Eric's comment above.

> how this will work in case there are multiple deployments

Until we have the ability to merge the documentation of all apps associated with a given host, I only see 2 options:

1. Only process the OpenAPI model for one web application per host, and fail to deploy others
2. Only process the OpenAPI model for one web application per host, and skip OpenAPI processing for the others.

> what the contents of the generated OpenAPI document will be with respect to the previously mentioned scenarios.

Without the ability to merged OpenAPI models, we are only able to process a single deployment into a single OpenAPI document.

Reply:
First of all, let me apologize for referencing hard requirement no.1 - while instead I meant to refer to this one:
```
* The `/openapi` endpoint shall be registered with the Undertow server/host associated with a deployment.
```
Which is no. 2, actually. So - again - sorry pointing you at the wrong item.

Now, about your last comment:

> This is mandated by the spec. See Eric's comment above.

Ok, so I am requesting for some more detail here - given previous examples involving different scenarios, like the ones you shared too (e.g.: virtual hosts, app root context etc.).

Sorry for repeating but the reader _here_ needs to be able and get all of the information in sufficient detail to determine how Wildfly would work with this feature enabled, and not going back to the spec in order to derive any missing information about the expected implementation.
This should be done by providing detailed description and [hopefully] examples in this very document, IMHO.
The above options about processing `a single deployment` and failing/skipping any OpenAPI generation for any deployment after the first one have to be documented, too.